connectors/flink/src/main/scala/io/delta/flink/internal/KernelDeltaLogDelegator.scala

```diff
@@ -0,0 +1,104 @@
+package io.delta.standalone.internal
+
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.Path
+
+import io.delta.kernel.{Table, TableNotFoundException}
+import io.delta.kernel.defaults.client.DefaultTableClient
+import io.delta.kernel.internal.{SnapshotImpl => SnapshotImplKernel, TableImpl}
+import io.delta.kernel.internal.fs.{Path => KernelPath}
+import io.delta.standalone.VersionLog
+import io.delta.standalone.actions.{CommitInfo => CommitInfoJ}
+import io.delta.standalone.internal.{SnapshotImpl => StandaloneSnapshotImpl}
+import io.delta.standalone.internal.util.{Clock, SystemClock}
+
+/**
+ *  Utility class to represent an "initial" snapshot as a kernel snapshot. This is equivalent to
+ *  InitialSnapshotImpl in standalone land.
+ */
+class InitialKernelSnapshotImpl(logPath: Path, dataPath: Path, tableClient: DefaultTableClient)
+    extends SnapshotImplKernel(
+      new KernelPath(logPath.toString()),
+      new KernelPath(dataPath.toString()),
+      -1,
+      io.delta.kernel.internal.snapshot.LogSegment.empty(new KernelPath(logPath.toString())),
+      tableClient,
+      -1)
+
+/**
+ * We want to be able to construct an OptimisticTransactionImpl that uses a delta log and a snapshot
+ * provided by the Delta Kernel. OptimisticTransactionImpl takes a DeltaLogImpl and SnapshotImpl
+ * internally, so we need classes that extend those, and this is the one for DeltaLogImpl. It
+ * provides features used by flink+startTransaction.
+ */
+class KernelDeltaLogDelegator(
+    tableClient: DefaultTableClient,
+    table: TableImpl,
+    hadoopConf: Configuration,
+    logPath: Path,
+    dataPath: Path,
+    clock: Clock)
+  extends DeltaLogImpl(hadoopConf, logPath, dataPath, clock) {
+
+  val standaloneDeltaLog = new DeltaLogImpl(hadoopConf, logPath, dataPath, clock)
+  var currKernelSnapshot: Option[KernelSnapshotDelegator] = None
+
+  override def snapshot(): StandaloneSnapshotImpl = { // but is actually a KernelSnapshotDelegator
+    if (currKernelSnapshot.isEmpty) { update() }
+    return currKernelSnapshot.get
+  }
+
+  override def update(): StandaloneSnapshotImpl = { // but is actually a KernelSnapshotDelegator
+    // get latest snapshot via kernel
+    val kernelSnapshot = try {
+      table.getLatestSnapshot(tableClient).asInstanceOf[SnapshotImplKernel]
+    } catch {
+      case e: TableNotFoundException =>
+        new InitialKernelSnapshotImpl(logPath, dataPath, tableClient)
+    }
+    currKernelSnapshot = Some(new KernelSnapshotDelegator(
+      kernelSnapshot,
+      hadoopConf,
+      logPath,
+      kernelSnapshot.getVersion(tableClient), // note: tableClient isn't used
+      this,
+      standaloneDeltaLog
+    ))
+    currKernelSnapshot.get
+  }
+
+  override def startTransaction(): io.delta.standalone.OptimisticTransaction = {
+    val snapshot = update()
+    new OptimisticTransactionImpl(this, snapshot)
```

Comment:
We want to be able to override `txnVersion`, to use Venki's `getRecentTransactionVersion` (err, hopefully renamed to `getLatestTransactionVersion` :p )

So, we will probably need to make a new class? or an inline anonymous class? just to override that 1 method

Reply:
It's already "there". Look at the definition of `transactions`, we override that. We'll have to return a special class that implements `Map` and waits to do any work until a particular appId is asked for, but that should be easy.

So OptimisticTransactionImpl does 
```scala
    snapshot.transactions.getOrElse(id, -1L)
```

and our `transactions` will return a Map that when you call `getOrElse` uses the kernel to get the version for that appId

Reply:
> We'll have to return a special class that implements Map and waits to do any work until a particular appId is asked for, but that should be easy.

SGTM

Reply:
mmmmmmm. actually .... `Map` has lots of APIs .... for example, `keySet` etc. We don't want to implement all of these APIs.

Think it's that terrible or that much extra work to just make a new class with 1 new API?